### ISSUE TITLE
Alternate Database Implementation

### ISSUE THREAD

**Issue by Noqomo** [alvyxaz/barebones-masterserver#5]
Hi all,

Anyone feel generous enough to share their code for changing the database solution to MySQL?

My C# skills are very rusty, and having trouble trying to get it to work.

Thanks!

**Comment by alvyxaz** [comment]
Hey!

If no one has a solution, I'll make one and share it within a few days - I guess there are more people who are interested in this ;)

**Comment by Noqomo** [comment]
Thanks mate!

**Comment by alvyxaz** [comment]
I was wondering if you need MySQL specifically, or do you think **MongoDB** would also work for you?

The reason why I'm asking is because I've switched most of my products to NoSQL databases, as I find them more convenient to work with.

It would also have a tool which you could use to insert/delete/edit entries.
(https://robomongo.org/)

**Comment by Noqomo** [comment]
Excellent I think that could work :)

**Comment by Noqomo** [comment]
Could you please elaborate on how I would implement MongoDB?

I'm still struggling with this part

**Comment by alvyxaz** [comment]
Hey,

Sure, I'll make a quick example this evening ;)

**Comment by Noqomo** [comment]
Thank you very much!

**Comment by alvyxaz** [comment]
Alright, so it seems like this won't be so easy to implement after all - mongodb drivers use assemblies that are not supported in the older mono environment, which unity uses. 

It's possible to work around this issue, but it got me thinking about something else:
https://forum.unity3d.com/threads/master-server-framework.430570/page-2#post-2849139

I'd love to hear your opinion on this. If you still want the framework to be built within Unity, I'll integrate the MongoDB solution into Unity.

If not, I was wondering if you'd be OK with waiting a few days, until I get the core of the framework moved outside of the Unity :)

**Comment by alvyxaz** [comment]
Just wanted to say that I've decided to keep the master server within Unity.

I'll post an example on how to use mongoDB in a few hours :)

**Comment by alvyxaz** [comment]
I took the source of the latest MongoDB C# drivers that still work on .NET 3.5, made some changes and stripped a few things that made it unable to work with Unity, so now it works. (had to throw out image serialization. It's still possible to do, but would need to be done

🔗 Custom MongoDB drivers: [Download](https://drive.google.com/open?id=0B87OgNU13WpjTXFzaGxVeU5MUWM)

⚠️ To make them work, you'll need to change your project File > Build Settings > Player settings > Other >API Compatibility Level to ".NET 2.0":

![](http://i.imgur.com/Opvmf6S.png)

I'm planning to include mongodb support officially in V1.03, but until then, if you're interested, here's a quick example on how to change the db implementation on the Auth module (other modules are similar)

1. First, you'll need to create your own implementation of the IAuthDatabase :
(I only tested registration, but everything else should also work)

``` C#
using System;
using MongoDB.Bson;
using MongoDB.Driver;
using MongoDB.Driver.Builders;

namespace Barebones.MasterServer
{
    /// <summary>
    /// MongoDB implementation of auth database
    /// http://mongodb.github.io/mongo-csharp-driver/1.11/getting_started/
    /// </summary>
    public class AuthDbMongo : IAuthDatabase
    {
        private MongoServer _server;
        private MongoDatabase _database;

        private MongoCollection<AccountDataMongo> _accountsCollection;
        private MongoCollection<PasswordResetData> _resetCodesCollection;
        private MongoCollection<EmailConfirmationData> _emailConfirmations;

        public const string EmailConfirmations = "emailConfirmations";

        public AuthDbMongo(string connectionString, string databaseName) 
            : this(new MongoClient(connectionString), databaseName)
        {
        }

        public AuthDbMongo(MongoClient client, string databaseName)
        {
            _server = client.GetServer();
            _database = _server.GetDatabase(databaseName);

            // Setup collections (a lot like tables in SQL databases)
            _accountsCollection = _database.GetCollection<AccountDataMongo>("accounts");
            _resetCodesCollection = _database.GetCollection<PasswordResetData>("pswResetCodes");
            _emailConfirmations = _database.GetCollection<EmailConfirmationData>("emailConfirmations");

            var options = IndexOptions.SetUnique(true);

            // Make usernames unique
            _accountsCollection.CreateIndex(IndexKeys<AccountDataMongo>.Ascending(e => e.Username), 
                new IndexOptionsBuilder().SetUnique(true));

            // Make emails unique
            _accountsCollection.CreateIndex(IndexKeys<AccountDataMongo>.Ascending(e => e.Email),
                new IndexOptionsBuilder().SetUnique(true));
        }

        public IAccountData CreateAccountObject()
        {
            return new AccountDataMongo();
        }

        public IAccountData GetAccount(string username)
        {
            var query = Query<AccountDataMongo>.EQ(e => e.Username, username);
            return _accountsCollection.FindOne(query);
        }

        public IAccountData GetAccountByToken(string token)
        {
            var query = Query<AccountDataMongo>.EQ(e => e.Token, token);
            return _accountsCollection.FindOne(query);
        }

        public IAccountData GetAccountByEmail(string email)
        {
            var query = Query<AccountDataMongo>.EQ(e => e.Email, email);
            return _accountsCollection.FindOne(query);
        }

        public void SavePasswordResetCode(IAccountData account, string code)
        {
            // Remove older entries
            var query = Query<PasswordResetData>.EQ(p => p.Email, account.Email);
            _resetCodesCollection.Remove(query);

            // Save a new one 
            _resetCodesCollection.Insert(new PasswordResetData()
            {
                Code = code,
                Email = account.Email
            });
        }

        public IPasswordResetData GetPasswordResetData(string email)
        {
            var query = Query<PasswordResetData>.EQ(p => p.Email, email);
            return _resetCodesCollection.FindOne(query);
        }

        public void SaveEmailConfirmationCode(string email, string code)
        {
            // Remove older entries
            var query = Query<EmailConfirmationData>.EQ(p => p.Email, email);
            _emailConfirmations.Remove(query);

            // Save a new one 
            _emailConfirmations.Insert(new EmailConfirmationData()
            {
                Code = code,
                Email = email
            });
        }

        public string GetEmailConfirmationCode(string email)
        {
            var query = Query<EmailConfirmationData>.EQ(e => e.Email, email);
            var result = _emailConfirmations.FindOne(query);
            return result != null ? result.Code : null;
        }

        public void UpdateAccount(IAccountData account)
        {
            _accountsCollection.Save(account as AccountDataMongo);
        }

        public void InsertNewAccount(IAccountData account)
        {
            _accountsCollection.Insert(account as AccountDataMongo);
        }

        public void InsertToken(IAccountData account, string token)
        {
            account.Token = token;
            UpdateAccount(account);
        }

        #region Concrete data schemes implementation

        private class EmailConfirmationData
        {
            // Necessary for each model
            public ObjectId Id { get; set; }

            public string Email { get; set; }
            public string Code { get; set; }
        }

        private class PasswordResetData : IPasswordResetData
        {
            // Necessary for each model
            public ObjectId Id { get; set; }

            public string Email { get; set; }
            public string Code { get; set; }
        }

        public class AccountDataMongo : IAccountData
        {
            // Necessary for each model
            public ObjectId Id { get; set; }

            public string Username { get; set; }
            public string Password { get; set; }
            public string Email { get; set; }
            public string Token { get; set; }
            public bool IsAdmin { get; private set; }
            public bool IsGuest { get; private set; }
            public bool IsEmailConfirmed { get; set; }
            public event Action<IAccountData> OnChange;

            public void MarkAsDirty()
            {
                if (OnChange != null)
                    OnChange.Invoke(this);
            }
        }

        #endregion
    }
}
```

2. Extend the `AuthModule` class with your custom class:

``` C#
using UnityEngine;

namespace Barebones.MasterServer
{
    public class MyAuthModule : AuthModule
    {
        [Header("MongoDB related")]
        public string ConnectionString = "mongodb://localhost";
        public string ConnectionStringArgName = "-bmMongo";
        public string DatabaseName = "masterServer";

        protected override IAuthDatabase SetupDatabase()
        {
            var arg = BmArgs.ExtractValue(ConnectionStringArgName);
            if (arg != null)
            {
                // If connection string was passed via arguments
                return new AuthDbMongo(arg, DatabaseName);
            }

            // Use default connection string
            return new AuthDbMongo(ConnectionString, DatabaseName);
        }
    }
}
```

3. Replace the `AuthModule` component with you custom one, in the Barebones>MasterServer>AuthModule` object

And that should be it. Let me know if you need any help ;)

**Comment by delmarle** [comment]
awesome for mongo DB in 1.03 ! on this wiki page: https://github.com/alvyxaz/barebones-masterserver/wiki/Player-Profiles

you talk about bidirectional, i think its kind of important in case you want to do things like these:
-regen currency every xx minute to all player from master server
-daily gift

if the player is connected to a game server then the bonus from "backend event" would be lost

**Comment by alvyxaz** [comment]
@delmarle 

This is a fair concern, and I also don't feel very comfortable with single-directional updates. I'll try to squeeze this into V1.03 too, because I'll need this for my little Christmas game, haha

**Comment by alvyxaz** [comment]
Included MongoDB in the V1.03 ;)
